substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/oldobject/OldObjectSampler.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2023, 2023, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2023, 2023, Red Hat Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.core.jfr.oldobject;
+
+import com.oracle.svm.core.Uninterruptible;
+
+import java.lang.ref.WeakReference;
+
+final class OldObjectSampler {
+    private final OldObjectArray samples;
+    private final OldObjectPriorityQueue queue;
+    private final OldObjectList list;
+    private final OldObjectEffects effects;
+    private long totalAllocated;
+
+    OldObjectSampler(int queueSize, OldObjectList list, OldObjectEffects effects) {
+        this.samples = new OldObjectArray(queueSize);
+        this.queue = new OldObjectPriorityQueue(this.samples);
+        this.list = list;
+        this.effects = effects;
+    }
+
+    @Uninterruptible(reason = "Accesses allocation profiler.")
+    void sample(WeakReference<Object> ref, long allocatedSize, int arrayLength) {
+        totalAllocated += allocatedSize;
+        long span = totalAllocated - queue.total();
+        if (queue.isFull()) {
+            if (queue.peek().span > span) {
+                // Sample will not fit, try to scavenge
+                // HotSpot code scavenges when it gets notified that a sample was GC'd,
```

Comment:
Theoretically, you could use a reference queue to get notified but I think that is too tricky to do from uninterruptible code. How does the notification work in HotSpot?